src/3DFactory.R

```diff
@@ -0,0 +1,138 @@
+# LT 17/05/2022
+
+# prototype: 3D Factory AND Adjudicator are in the same source file to avoid
+# implementation of the interface between them
+
+# 3D Factory component
+
+# the 3D Factory generates a list of 3D regions and calculates their intolerance score
+
+# First prototype
+# creates 3D regions of variable length using upper bound of CI
+# each region is built by growing a sphere centered on the carbon-alpha of
+# a specific residue until the whole protein is included
+# a criteria (based on the upper bound) is used to decide the length of the region
+
+# file names and parameters are hard-coded for this prototype
+
+source("constraint_utils.R")
+
+## calculate upper bound of the o/e confidence interval
+get_oe_ci_up <- function(observed, expected, alpha = 0.05) {
+  return(qchisq((1 - alpha / 2), 2 * (observed + 1)) / 2 / expected)
+}
+
+## calculate intolerance metric of a region
+get_region_metric <- function(i_region, alpha, regions, oe_data) {
+  # indices of the residues belonging to this region
+  this_region <- regions[seq_len(i_region)]

```

Comment:
Not a blocker on this PR, but it would be interesting to think about how you might expand the region creation step to test spheres that can't be built through this iterative add-the-nearest-residue approach. 

I.e. as long as you do this exhaustive search (adding a single residue at a time, testing every single residue as the starting point for region formation), it should be great for capturing conserved regions if they are centred on a given residue. But for things like binding pockets, you could imagine the centre point for the conservation signal would actually be _between_ residues. Forming spheres around any of the relevant pocket residues may still capture signal, but it's likely to be diluted by non-binding-pocket residues that are in more immediate proximity to the starting point.

Reply:
Yes it is on my list. Technically it is simple to make the same calculations using an arbitrary centre for the sphere: just use the 3d coordinates of the alpha-carbons and calculate the distance between the centre of the sphere and each residues. It is computationally cheap.
The real question therefore is how to choose the position of the centre of the sphere. I have so far considered three strategies:
- brute force. Generate a large number of spheres with random positions, keep the best ones. Computationally expensive.
- optimization. Starting form a candidate sphere centre, optimize its position using gradient descent. This would require the intolerance metric to be continuous and derivable over space with could be achieved, for example, by using a Gaussian kernel
- weighted average / barycentre. Find intolerant spheres close to each other and use the barycentre of their centres
All this needs further thought but before complicating the approach I want to collect hard evidence that it is needed. This is why we have chosen a couple of proteins with constrained beta-barrels for the validation phase

Reply:
Regarding [this comment](https://github.com/populationgenomics/structural-constraint/pull/28#discussion_r884199637)
Agreed, this was  temporary for the demo.
Regarding formatting: we made the decision of using `styler` after I has started coding. In the future the formatting will be right from the beginning
Regarding test data: my view is that the results of analyses should not even be on Github, or if it is in a separate directory examples.